tests/test_layabout.py

```diff
@@ -1 +1,327 @@
+    with pytest.raises(FailedConnection) as exc:
+        layabout.run(
+            token=TOKEN,
+            retries=1,
+            until=lambda e: False
+        )
+
+    assert str(exc.value) == 'Failed to reconnect to the Slack API'
+
+
+def test_layabout_can_reuse_an_existing_client_to_reconnect(monkeypatch):
+    """
+    Test that layabout can reuse an existing SlackClient instance to reconnect
+    to the Slack API rather than needlessly instantiating a new one on each
+    reconnection attempt.
+    """
+    layabout = Layabout(name='test')
+
+    connections = [
+        False,  # Fail the initial connection (_connect).
+        False,  # Fail with the first reconnection attempt (_reconnect).
+        True,  # Succeed with the second reconnection attempt (_reconnect).
+    ]
+
+    rtm_connect = MagicMock(side_effect=connections)
+    slack = MagicMock(rtm_connect=rtm_connect)
+    SlackClient = MagicMock(return_value=slack)
+    monkeypatch.setattr('layabout.SlackClient', SlackClient)
+
+    # Retry connecting twice to verify reconnection logic was evaluated.
+    # until will exit early just to be safe.
+    layabout.run(
+        token=TOKEN,
+        retries=2,
+        backoff=lambda r: 0,
+        until=lambda e: False
+    )
+
+    # Make sure the SlackClient was only instantiated once so we know that we
+    # reused the existing instance.
+    SlackClient.assert_called_once_with(token=TOKEN)
+
+
+def test_layabout_can_continue_after_successful_reconnection(monkeypatch):
+    """
+    Test that layabout can continue to handle events after successfully
+    reconnecting to the Slack API.
+    """
+    layabout = Layabout(name='test')
+
+    connections = [
+        True,  # Succeed with the first connection (_connect).
+        True,  # Succeed later with a reconnection (_reconnect).
+    ]
+    reads = [
+        TimeoutError,  # Raise an exception on the first read.
+        [],  # Return empty events on the second read (after reconnection).
+    ]
+
+    rtm_connect = MagicMock(side_effect=connections)
+    rtm_read = MagicMock(side_effect=reads)
+    slack = MagicMock(rtm_connect=rtm_connect, rtm_read=rtm_read)
+
+    SlackClient = MagicMock(return_value=slack)
+    monkeypatch.setattr('layabout.SlackClient', SlackClient)
+
+    layabout.run(
+        token=TOKEN,
+        retries=1,
+        backoff=lambda r: 0,
+        until=lambda e: False
+    )
+
+
+def test_layabout_can_handle_events(monkeypatch):
+    """
+    Test that layabout can handle events appropriately given multiple event
+    handlers and even * event handlers.
+    """
+    events = [
+        dict(type='hello'),
+        dict(type='goodbye'),
+    ]
+    layabout = Layabout(name='test')
+
+    connections = [
+        True,  # Succeed with the first connection (_connect).
+    ]
+
+    rtm_connect = MagicMock(side_effect=connections)
+    rtm_read = MagicMock(return_value=events)
+    slack = MagicMock(rtm_connect=rtm_connect, rtm_read=rtm_read)
```

Comment:
Aye 👍 for re-usable testing resources.

Reply:
I see what you guys are saying and I agree this can (and should) be simplified, but I don't see an easy way to do it yet. There are a lot of similarities, but some differences between setups and I need them to be configurable, so a `pytest.fixture` doesn't seem quite right here. Perhaps I can craft an auxiliary function like
```
def mock_slack_client(connections, events):
    ...
```
and use it like
```python
SlackClient = mock_slack_client([False, True], [dict(...), dict(...)])
```

Reply:
I'm a fan of using your codes setup functions to build the required objects.

Reply:
I don't follow. In this case, I don't have a setup function that builds a fake `SlackClient`...

Reply:
Your concept of a `mock_slack_client` helper is pretty much what I was thinking, though I would use keyword arguments so it's clearer what it's doing at the call site.  The idea is just to condense the setup but retain the useful bits of information that make each test setup unique.

Reply:
I'm not really sure what I meant when I wrote that. =P

I think I was thinking about stating that configurable objects should have instantiation done in the code base and you should use the codes setup functions but that doesn't make any sense in the context of mocks.